ovos_plugin_manager/templates/tts.py

```diff
@@ -412,58 +415,115 @@ def execute(self, sentence, ident=None, listen=False, **kwargs):
         chunks = [(chunks[i], listen if i == len(chunks) - 1 else False)
                   for i in range(len(chunks))]
         self.handle_metric({"metric_type": "tts.preprocessed",
                             "n_chunks": len(chunks)})
-        for sentence, l in chunks:
-            key = str(hashlib.md5(
-                sentence.encode('utf-8', 'ignore')).hexdigest())
-            wav_file = os.path.join(self.cache_dir, key + '.' + self.audio_ext)
 
-            if os.path.exists(wav_file):
-                LOG.debug("TTS cache hit")
-                phonemes = self.load_phonemes(key)
-            else:
-                self.handle_metric({"metric_type": "tts.synth.start"})
-                lang = kwargs.get("lang")
-                if not lang and kwargs.get("message"):
-                    # some HolmesV derivatives accept a message object
-                    try:
-                        lang = kwargs["message"].data.get("lang") or \
-                               kwargs["message"].context.get("lang")
-                    except:  # not a mycroft message object
-                        pass
-                lang = lang or self.lang
-                # check the signature to either pass lang or not
-                if len(signature(self.get_tts).parameters) == 3:
-                    wav_file, phonemes = self.get_tts(sentence, wav_file,
-                                                      lang=lang)
-                else:
-                    wav_file, phonemes = self.get_tts(sentence, wav_file)
-                self.handle_metric({"metric_type": "tts.synth.finished"})
-                if phonemes:
-                    self.save_phonemes(key, phonemes)
-                else:
-                    try:
-                        # TODO, debug why phonemes fail ?
-                        phonemes = get_phonemes(sentence)
-                        self.handle_metric({"metric_type": "tts.phonemes.guess"})
-                    except (ImportError, FailedToGuessPhonemes):
-                        pass
-            vis = self.viseme(phonemes) if phonemes else None
-            self.queue.put((self.audio_ext, wav_file, vis, ident, l))
+        # synth -> queue for playback
+        for sentence, l in chunks:
+            sentence_hash = hash_sentence(sentence)
+            if sentence_hash in self.cache:  # load from cache
+                audio_file, phonemes = self._get_from_cache(sentence, sentence_hash)
+            else:  # synth + cache
+                audio_file, phonemes = self._synth(sentence, sentence_hash, **kwargs)
+
+            viseme = self.viseme(phonemes) if phonemes else None
+            audio_ext = self._determine_ext(audio_file)
+            self.queue.put(
+                (audio_ext, str(audio_file), viseme, ident, l)
+            )
             self.handle_metric({"metric_type": "tts.queued"})
 
+    def _determine_ext(self, audio_file):
+        # determine audio_ext on the fly
```

Comment:
A potentially better approach here is to read it from file headers in case the plugin incorrectly specifies (i.e. a remote TTS implementation changes)

Reply:
i dont think plugins should specify it at all to be honest, but that needs a larger refactor. and mycroft is also talking about the same so lets wait and see.... mycroft intends to deprecate allowing TTS to return file path, so there will be breakage either way....

the aim here is to allow plugins to return different extensions per request, i had this happening when i added a permanent cache to polly that was in .wav format, but the plugin returns mp3.